Update market calendar

Stryder-Git opened the pull request:
@rsheftel, I am posting this PR as a draft because new functionality still needs more tests and proper documentation. But in the meantime, you and anyone who is interested already have the chance to give feedback or make additional requests. 

Also, I apologize for the large diffs, it's going to be hard to split this into multiple PRs since most changes depend on each other, but if I will if you want me to. 

This is a brief overview of what this PR would change:

* MarketCalendar
	* uses the `regular_market_times` dictionary with market times (open/close/etc.) in a similar way to the exchange_calendars package, which will allow all MarketCalendar children to handle changing market times with ease.
	* has a few extra helper methods, allowing to customize market_times, check what the regular market_time was on a certain date, and if any market times have been discontinued (e.g. XKRX stopped having breaks in 2000).
	* .schedule will adjust to all market time information that the instance has available and is very customizable, while being fully compatible with the old version

* NYSEExchangeCalendar

	* the overarching changes in MarketCalendar allowed this class to be simplified and is a good example of how to handle special cases (e.g. Saturday being a trading day) with the new setup
	* removed all the string conversions and list comprehensions in special_opens/closes
	* for other changes please consult ...

* holidays_nyse.py
	* please consult ...

* class_registry.py
	* simplified the registration/lookup of classes a bit
	* added the ProtectedDict class to make sure regular_market_times are adjusted in the right way
	* added some additional set up requirements to RegistryMeta._ _ init _ _

* exchange_calendars_mirror.py
	* fit the data on market times to the new format in MarketCalendar

* exchange_calendars_*
	* changed the implementation of the MarketCalendar children to fit the new format

* various tweaks in various tests

This PR would be changing quite a lot, so I hope you are alright with this, @rsheftel.
Any additional suggestions/requests/concerns from anyone who reads this are more than welcome.

rsheftel commented:
@Stryder-Git this is great work, thank you very much. I will try to take a look over the coming days, but at first glance everything looks to be a solidly backwards compatible improvement.

Stryder-Git commented:
Hi @rsheftel,

One point I forgot to mention:
	
* I completely removed the `open_time_default` and `close_time_default` properties since they were not fulfilling a purpose anymore and are not represented in any tests. Checking if a market_time is customized can be done through `.is_custom(market_time)`, which is safer than comparing .open_time to .open_time_default. Let me know if you would like me to add them again.

Something I just finished, regarding `exchange_calendars_mirror.TradingCalendar`:
	
* offset times are also incorporated now, which means that all the variable open/close/break times AND their offsets are fully integrated into this package's class

* The large calculations in `__init__`s of exchange_calendar classes are fully bypassed when using a mirrored calendar, but all the data on market_times is still available, which brings the load time of one of these calendars *from 300-500ms to less than 0.1 ms*, while retaining the information of exchange_calendars and adding the new flexibility of pandas_market_calendars.

* Accessing the mirrored ExchangeCalendar can be done through the `.ec` property, which will run the aforementioned `__init__` method only the first time it is used, and provides access to the full-fledged exchange_calendars calendar.

* I renamed all references to exchange_calendars classes to `ec` rather than `tc` since I thought that would be a bit more intuitive, let me know if you'd like me to change it back.

Either later today or tomorrow, I will probably add to the examples/usage.ipynb file and link to it here, so you can have a look at the new functionality that I am creating. 

Let me know if you disagree with any of these changes or if you have any additional suggestions, thanks!

maread99 commented:
Hi @Stryder-Git, this PR looks interesting. I collaborate on `exchange_calendars` and will certainly be keeping an eye on this PR to see if there's anything that can be similarly improved over there.

I see you seem to be familiar with the `exchange_calendars` implementation. I would certainly be interested in hearing any suggestions that might occur to you as a result of your work on this PR!

Cheers.

rsheftel commented:
@Stryder-Git this all looks great. Thanks again for the hard work. I will take a look at the examples when you commit and then assuming everything ties out we can merge this PR.

Stryder-Git commented:
@maread99, I will get back to you about your comments here and in other places.

I just commited again with the new version of examples/usage.ipynb, pretty much all changes are between the 'Customizations' and 'Helpers' headings, but I probably missed some stuff or need to clarify some things. I'm also not sure if all of that should be in the usage notebook but I will figure out how to document this better after I have written more tests.

Until then, feel free to criticize and suggest, I am looking forward to all ideas and opinions!

rsheftel commented:
It appears this branch passes all tests and looks correct. Are you comfortable with me merging it?

Stryder-Git commented:
Sorry for not commenting but yes, from a technical standpoint, and if you are alright with leaving the docs as they are, this PR is good to go. I might just have another PR coming in the next week or two to improve the docs and possibly make a few optimization/usability tweaks. I hope you like the changes!

rsheftel commented:
This is great. I will wait for those next changes from you and then merge it into the master.
